### PR TITLE
[beta] Backport accepted PRs to 1.21

### PR DESCRIPTION
**Pull request by dtolnay**
Backport of:

- ~don't suggest placing `use` statements into expanded code https://github.com/rust-lang/rust/pull/44215~
- stabilize tcpstream_connect_timeout https://github.com/rust-lang/rust/pull/44563
- stabilized iterator_for_each https://github.com/rust-lang/rust/pull/44567
- travis: Move sccache to the us-west-1 region https://github.com/rust-lang/rust/pull/44574
- stabilized ord_max_min https://github.com/rust-lang/rust/pull/44593
- stabilized compiler_fences https://github.com/rust-lang/rust/pull/44595
- ci: Upload/download from a new S3 bucket https://github.com/rust-lang/rust/pull/44617
- stabilized needs_drop https://github.com/rust-lang/rust/pull/44639
- Stabilized vec_splice and modified splice tracking issue https://github.com/rust-lang/rust/pull/44640
- Backport libs stabilizations to 1.21 beta https://github.com/rust-lang/rust/pull/44824

### CODE REVIEW

File: src/liballoc/string.rs [hunk 1]
```diff
@@ -1420,7 +1420,7 @@ impl String {
     /// assert_eq!(t, "α is alpha, ");
     /// assert_eq!(s, "Α is capital alpha; β is beta");
     /// ```
-    #[unstable(feature = "splice", reason = "recently added", issue = "32310")]
+    #[unstable(feature = "splice", reason = "recently added", issue = "44643")]
```
Comment (kennytm) [inline src/liballoc/string.rs]: Please rename the feature.

```
[00:04:19] tidy error: /checkout/src/liballoc/string.rs:2253: mismatches to previous in: ["tracking issue"]
```

**Review by alexcrichton — APPROVED** [review]
r=me wit hthe tidy error fixed,

thanks @dtolnay!

**Comment by alexcrichton** [comment]
@bors: r-

backport fails to compile?

**Comment by dtolnay** [comment]
I removed the backport of #44215 from this PR, will leave that one to someone else.

<details>
<summary>(expand) Here is the merge I botched.</summary>

```rust
            match item.node {
                ItemKind::Use(..) => {
                    // don't suggest placing a use before the prelude
                    // import or other generated ones
<<<<<<< HEAD
                    if item.span == DUMMY_SP {
                        let mut span = item.span;
                        span.hi = span.lo;
                        self.span = Some(span);
||||||| parent of 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                    if item.span == DUMMY_SP {
                        self.span = Some(item.span.with_hi(item.span.lo()));
=======
                    if item.span.ctxt().outer().expn_info().is_none() {
                        self.span = Some(item.span.with_hi(item.span.lo()));
>>>>>>> 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                        self.found_use = true;
                        return;
                    }
                },
                // don't place use before extern crate
                ItemKind::ExternCrate(_) => {}
                // but place them before the first other item
                _ => if self.span.map_or(true, |span| item.span < span ) {
<<<<<<< HEAD
                    let mut span = item.span;
                    span.hi = span.lo;
                    self.span = Some(span);
||||||| parent of 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                    self.span = Some(item.span.with_hi(item.span.lo()));
=======
                    if item.span.ctxt().outer().expn_info().is_none() {
                        // don't insert between attributes and an item
                        if item.attrs.is_empty() {
                            self.span = Some(item.span.with_hi(item.span.lo()));
                        } else {
                            // find the first attribute on the item
                            for attr in &item.attrs {
                                if self.span.map_or(true, |span| attr.span < span) {
                                    self.span = Some(attr.span.with_hi(attr.span.lo()));
                                }
                            }
                        }
                    }
>>>>>>> 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                },
```
</details>

**Pull request merged**